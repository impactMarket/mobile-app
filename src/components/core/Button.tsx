import { iptcColors } from 'styles/index';
import React, { Component } from 'react';
import { StyleProp, TextStyle, ViewStyle } from 'react-native';
import { Button as RNPButton, Text } from 'react-native-paper';

interface IButtonProps {
    modeType: 'green' | 'gray' | 'default';
    bold?: boolean;
    icon?: string;
    children?: any;
    loading?: boolean;
    disabled?: boolean;
    onPress?: () => void;
    style?: StyleProp<ViewStyle>;
    contentStyle?: StyleProp<ViewStyle>;
    labelStyle?: StyleProp<TextStyle>;
}
interface IButtonState {}

export default class Button extends Component<IButtonProps, IButtonState> {
    constructor(props: any) {
        super(props);
    }

    render() {
        const { style, disabled, children, modeType, bold } = this.props;
        let buttonStyle: StyleProp<ViewStyle> = {
            elevation: 0,
            height: 42,
            justifyContent: 'center',
            // borderRadius: 4,
        };
        if (!disabled) {
            if (modeType === 'green') {
                buttonStyle = {
                    ...buttonStyle,
                    backgroundColor: iptcColors.greenishTeal,
                };
            } else if (modeType === 'gray') {
                buttonStyle = {
                    ...buttonStyle,
                    backgroundColor: iptcColors.softGray,
                };
            } else {
                buttonStyle = {
                    ...buttonStyle,
                    backgroundColor: iptcColors.softBlue,
                };
            }
        }
        if (this.props.style) {
            buttonStyle = { ...buttonStyle, ...(style as any) };
        }
        return (
            <RNPButton
                mode="contained"
                uppercase={false}
                {...this.props}
                style={buttonStyle}
                labelStyle={{
                    color: modeType === 'gray' ? '#32325D' : 'white',
                    fontFamily: bold ? 'Gelion-Bold' : 'Gelion-Regular',
                    fontSize: 15,
                    lineHeight: 17.58,
                    ...(this.props.labelStyle as any),
                }}
            >
                {children}
            </RNPButton>
        );
    }
}
