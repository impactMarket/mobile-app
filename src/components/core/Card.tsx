import { Card as RNPCard } from 'react-native-paper';
import styled from 'styled-components/native';

const Card = styled(RNPCard)`
    border-radius: 8px;
    elevation: 6;
    box-shadow: 0px 2px 14px #e1e4e7;
`;

export default Card;
