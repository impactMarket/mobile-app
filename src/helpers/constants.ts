export enum Screens {
    CommunityDetails = 'CommunityDetails',
    FAQ = 'FAQ',
    Welcome = 'Welcome',
    CreateCommunity = 'CreateCommunity',
    WaitingTx = 'WaitingTx',
    Stories = 'Stories',
    NewStory = 'NewStory',
    StoriesCarousel = 'StoriesCarousel',
    Carousel = 'Carousel',
    //
    ClaimExplained = 'ClaimExplained',
    AnonymousReport = 'AnonymousReport',
    AddedBeneficiary = 'AddedBeneficiary',
    RemovedBeneficiary = 'RemovedBeneficiary',
    AddBeneficiary = 'AddBeneficiary',
    AddManager = 'AddManager',
    AddedManager = 'AddedManager',
    //
    Beneficiary = 'Beneficiary',
    WelcomeRulesScreen = 'WelcomeRulesScreen',
    CommunityManager = 'CommunityManager',
    Communities = 'Communities',
    Profile = 'Profile',
    Auth = 'Auth',
}

// async storage
export const STORAGE_USER_ADDRESS = '@wallet:address';
export const STORAGE_USER_PHONE_NUMBER = '@wallet:phonenumber';
export const STORAGE_USER_AUTH_TOKEN = '@user:authtoken';

// redux
// export const INIT_USER = 'INIT_USER';
export const SET_USER_CELO_INFO = 'SET_USER_CELO_INFO';
export const SET_USER_INFO = 'SET_USER_INFO';
export const SET_USER_EXCHANGE_RATE = 'SET_USER_EXCHANGE_RATE';
export const SET_USER_WALLET_BALANCE = 'SET_USER_WALLET_BALANCE';
export const SET_CELO_KIT = 'SET_CELO_KIT';
export const SET_COMMUNITY_CONTRACT = 'SET_COMMUNITY_CONTRACT';
export const SET_COMMUNITY = 'SET_COMMUNITY';
export const SET_USER_IS_BENEFICIARY = 'SET_USER_IS_BENEFICIARY';
export const SET_USER_IS_BLOCKED = 'SET_USER_IS_BLOCKED';
export const SET_USER_IS_SUSPECT = 'SET_USER_IS_SUSPECT';
export const SET_USER_IS_COMMUNITY_MANAGER = 'SET_USER_IS_COMMUNITY_MANAGER';
export const RESET_USER_APP = 'RESET_USER_APP';
export const RESET_NETWORK_APP = 'RESET_NETWORK_APP';
export const SET_PUSH_NOTIFICATION_TOKEN = 'SET_PUSH_NOTIFICATION_TOKEN';
export const SET_AUTH_TOKEN = 'SET_AUTH_TOKEN';
export const SET_USER_LANGUAGE = 'SET_USER_LANGUAGE';
export const SET_EXCHANGE_RATES = 'SET_EXCHANGE_RATES';
export const CONSENT_ANALYTICS = 'CONSENT_ANALYTICS';
export const SHOW_REPORT_CARD = 'SHOW_REPORT_CARD';

export const SET_USER_AUTH_REQUEST = 'SET_USER_AUTH_REQUEST';
export const SET_USER_AUTH_SUCCESS = 'SET_USER_AUTH_SUCCESS';
export const SET_USER_AUTH_FAILURE = 'SET_USER_AUTH_FAILURE';

export const SET_APP_SUSPECT_WRONG_DATETIME = 'app/suspectWrongDateTime';
export const SET_APP_FROM_WELCOME_SCREEN = 'app/fromWelcomeScreen';
export const SET_APP_BENEFICIARY_HAS_ACCEPTED_TERMS =
    'app/hasBeneficiaryAcceptedRulesAlready';
export const SET_APP_MANAGER_HAS_ACCEPTED_TERMS =
    'app/hasManagerAcceptedRulesAlready';

export enum appAction {
    SET_PUSH_NOTIFICATION_LISTENERS = 'app/setPushNotificationListeners',
}

export enum modalDonateAction {
    OPEN = 'donateModal/open',
    GO_TO_CONFIRM_DONATE = 'donateModal/goToConfirm',
    GO_TO_ERROR_DONATE = 'donateModal/goToError',
    GO_BACK_TO_DONATE = 'donateModal/goBackToDonate',
    IN_PROGRESS = 'donateModal/inProgress',
    CLOSE = 'donateModal/close',
}

export enum apiImageTargets {
    COVER = '/community/picture',
    PROFILE = '/user/picture',
}

export enum imageTargets {
    COVER = 'COVER',
    PROFILE = 'PROFILE',
}

export enum storiesAction {
    INIT_REQUEST = 'stories/init_request',
    INIT_SUCCESS = 'stories/init_success',
    INIT_FAILURE = 'stories/init_failure',
    CONCAT = 'stories/concat',
    USER_STORIES_REQUEST = 'stories/user_request',
    USER_STORIES_SUCCESS = 'stories/user_success',
    USER_STORIES_FAILURE = 'stories/user_failure',
}

export const SET_VIEW_MANAGER_DETAILS = 'view/managerDetails';

export enum celoNetwork {
    noAddress = '0x0000000000000000000000000000000000000000',
}
