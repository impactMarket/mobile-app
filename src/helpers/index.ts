import { ContractKit } from '@celo/contractkit';
import i18n from 'assets/i18n';
import BigNumber from 'bignumber.js';
import { ethers } from 'ethers';
import Api from 'services/api';

import config from '../../config';
import CommunityContractABI from '../contracts/CommunityABI.json';
import ImpactMarketContractABI from '../contracts/ImpactMarketABI.json';
import { ImpactMarketInstance } from '../contracts/types/truffle-contracts';
import {
    setCommunityContract,
    setUserIsBeneficiary,
    setUserIsCommunityManager,
    setImpactMarketContract,
    setCommunity,
} from './redux/actions/ReduxActions';
import { ICommunityInfo, IUser, IUserState, IUserInfo } from './types';

const usergetUserAvatars = [
    [
        require('assets/images/avatar/round/avatar1.png'),
        require('assets/images/avatar/round/avatar2.png'),
        require('assets/images/avatar/round/avatar3.png'),
        require('assets/images/avatar/round/avatar4.png'),
        require('assets/images/avatar/round/avatar5.png'),
        require('assets/images/avatar/round/avatar6.png'),
        require('assets/images/avatar/round/avatar7.png'),
        require('assets/images/avatar/round/avatar8.png'),
    ],
    [
        require('assets/images/avatar/square/avatar1.png'),
        require('assets/images/avatar/square/avatar2.png'),
        require('assets/images/avatar/square/avatar3.png'),
        require('assets/images/avatar/square/avatar4.png'),
        require('assets/images/avatar/square/avatar5.png'),
        require('assets/images/avatar/square/avatar6.png'),
        require('assets/images/avatar/square/avatar7.png'),
        require('assets/images/avatar/square/avatar8.png'),
    ],
];

export function getAvatarFromId(avatarId: number, big: boolean = false) {
    return usergetUserAvatars[big ? 0 : 1][avatarId];
}

export function getUserAvatar(user: IUserInfo, big: boolean = false) {
    if (user.avatar.length < 3) {
        return usergetUserAvatars[big ? 0 : 1][parseInt(user.avatar) - 1];
    }
    return { uri: user.avatar };
}

export function getUserCurrencySymbol(user: IUserInfo) {
    switch (user.currency.toUpperCase()) {
        case 'EUR':
            return 'â‚¬';
        default:
            return '$';
    }
}

export function amountToUserCurrency(
    amount: BigNumber | string,
    user: IUserInfo
) {
    const exchangeRate = user.exchangeRate;
    const bgn = new BigNumber(amount).multipliedBy(exchangeRate);
    const result = humanifyNumber(bgn);
    return result;
}

export function claimFrequencyToText(frequency: BigNumber | string): string {
    const f = new BigNumber(frequency);
    if (f.eq(3601)) return i18n.t('hour');
    if (f.eq(86400)) return i18n.t('day');
    if (f.eq(604800)) return i18n.t('week');
    return 'month';
}

// cUSD has 18 zeros!
export function humanifyNumber(inputNumber: BigNumber | string): number {
    const decimals = new BigNumber(10).pow(config.cUSDDecimals);
    return parseFloat(new BigNumber(inputNumber).div(decimals).toFixed(2));
}

export function calculateCommunityProgress(
    toCalculte: string /*'raised' | 'claimed'*/,
    community: ICommunityInfo
): number {
    const m = new BigNumber(community.vars._maxClaim).multipliedBy(
        community.beneficiaries.added.length +
            community.beneficiaries.removed.length
    );
    const result = new BigNumber(
        toCalculte === 'raised' ? community.totalRaised : community.totalClaimed
    ).div(m.eq(0) ? 1 : m);
    return parseFloat(result.toFixed(2));
}

export function getCountryFromPhoneNumber(phoneNumber: string) {
    if (phoneNumber.slice(0, 4) === '+351') {
        return 'ðŸ‡µðŸ‡¹ Portugal';
    }
}

export var iptcColors = {
    greenishTeal: '#2dce89',
    softBlue: '#5e72e4',
};

export async function loadContracts(
    address: string,
    kit: ContractKit,
    store: any
) {
    const fSetCommunity = (c: ICommunityInfo) => {
        // c.contractAddress can be null if community approval is still pending
        const communityContract = new kit.web3.eth.Contract(
            CommunityContractABI as any,
            c.contractAddress
        );
        store.dispatch(setCommunity(c));
        store.dispatch(setCommunityContract(communityContract));
    };
    const provider = new ethers.providers.Web3Provider(
        kit.web3.currentProvider as any
    );
    const impactMarketContract = new ethers.Contract(
        config.impactMarketContractAddress,
        ImpactMarketContractABI,
        provider
    ) as ethers.Contract & ImpactMarketInstance;
    store.dispatch(setImpactMarketContract(impactMarketContract));

    const isBeneficiary = await Api.findComunityToBeneficicary(address);
    if (isBeneficiary !== undefined) {
        store.dispatch(setUserIsBeneficiary(true));
        fSetCommunity(isBeneficiary);
        return;
    }
    const isManager = await Api.findComunityToManager(address);
    if (isManager !== undefined) {
        store.dispatch(setUserIsCommunityManager(true));
        fSetCommunity(isManager);
    }
}

export async function updateCommunityInfo(address: string, store: any) {
    const isBeneficiary = await Api.findComunityToBeneficicary(address);
    if (isBeneficiary !== undefined) {
        store.dispatch(setCommunity(isBeneficiary));
        return;
    }
    const isManager = await Api.findComunityToManager(address);
    if (isManager !== undefined) {
        store.dispatch(setCommunity(isManager));
    }
}

export function validateEmail(email: string) {
    const emailRegex = /^[-!#$%&'*+\/0-9=?A-Z^_a-z{|}~](\.?[-!#$%&'*+\/0-9=?A-Z^_a-z`{|}~])*@[a-zA-Z0-9](-*\.?[a-zA-Z0-9])*\.[a-zA-Z](-?[a-zA-Z0-9])+$/;
    if (!email) return false;

    if (email.length > 254) return false;

    var valid = emailRegex.test(email);
    if (!valid) return false;

    // Further checking of some things regex can't handle
    var parts = email.split('@');
    if (parts[0].length > 64) return false;

    var domainParts = parts[1].split('.');
    if (
        domainParts.some(function (part) {
            return part.length > 63;
        })
    )
        return false;

    return true;
}
