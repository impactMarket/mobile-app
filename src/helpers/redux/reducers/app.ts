import {
    appAction,
    SET_APP_FROM_WELCOME_SCREEN,
    SET_APP_SUSPECT_WRONG_DATETIME,
    SET_CELO_KIT,
    SET_EXCHANGE_RATES,
} from 'helpers/constants';
import { AppActionTypes } from 'helpers/types/redux';
import { IAppState } from 'helpers/types/state';

const INITIAL_STATE_APP: IAppState = {
    kit: undefined as any,
    exchangeRates: undefined as any, // save exhangeRates on load
    suspectWrongDateTime: false,
    timeDiff: 0,
    fromWelcomeScreen: '',
};

export const appReducer = (
    state = INITIAL_STATE_APP,
    action: AppActionTypes
) => {
    switch (action.type) {
        case SET_CELO_KIT:
            return { ...state, kit: action.payload };
        case SET_EXCHANGE_RATES:
            return { ...state, exchangeRates: action.payload };
        case SET_APP_SUSPECT_WRONG_DATETIME:
            return {
                ...state,
                suspectWrongDateTime: action.payload.suspect,
                timeDiff: action.payload.timeDiff,
            };
        case SET_APP_FROM_WELCOME_SCREEN:
            return {
                ...state,
                fromWelcomeScreen: action.payload,
            };
        case appAction.SET_PUSH_NOTIFICATION_LISTENERS:
            return {
                ...state,
                notificationsListeners: action.payload,
            };
        default:
            return state;
    }
};
