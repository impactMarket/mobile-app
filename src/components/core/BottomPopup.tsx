import CloseStorySvg from 'components/svg/CloseStorySvg';
import React, { ReactElement } from 'react';
import {
    GestureResponderEvent,
    View,
    TextStyle,
    StyleProp,
} from 'react-native';
import { BottomSheet } from 'react-native-btr';
import { Headline } from 'react-native-paper';

import i18n from '../../assets/i18n';
import Card from './Card';

interface IBottomPopupProps {
    children: ReactElement;
    isVisible: boolean;
    hasCloseBtn?: boolean;
    title?: string;
    titleStyle?: StyleProp<TextStyle>;
    toggleVisibility?: () => void;
}
export default class BottomPopup extends React.Component<
    IBottomPopupProps,
    object
> {
    render() {
        const {
            children,
            isVisible,
            toggleVisibility,
            title,
            titleStyle,
            hasCloseBtn,
        } = this.props;
        return (
            <BottomSheet
                visible={isVisible}
                onBackButtonPress={toggleVisibility}
                onBackdropPress={toggleVisibility}
            >
                <Card
                    style={{
                        borderBottomEndRadius: 0,
                        borderBottomStartRadius: 0,
                    }}
                >
                    <Card.Content>
                        <View
                            style={{
                                flexDirection: 'row',
                                justifyContent: 'space-between',
                            }}
                        >
                            {title && (
                                <Headline
                                    style={
                                        titleStyle
                                            ? [
                                                  {
                                                      fontWeight: '800',
                                                      fontFamily:
                                                          'Manrope-Bold',
                                                  },
                                                  titleStyle,
                                              ]
                                            : {
                                                  fontWeight: '500',
                                                  textAlign: 'center',
                                                  fontFamily: 'Manrope-Bold',
                                              }
                                    }
                                >
                                    {i18n.t(title)}
                                </Headline>
                            )}
                            {hasCloseBtn && toggleVisibility && (
                                <CloseStorySvg
                                    onPress={() => toggleVisibility()}
                                />
                            )}
                        </View>
                        <View
                            style={{
                                alignItems: 'center',
                                marginTop: 20,
                                marginBottom: 10,
                            }}
                        >
                            {children}
                        </View>
                    </Card.Content>
                </Card>
            </BottomSheet>
        );
    }
}
