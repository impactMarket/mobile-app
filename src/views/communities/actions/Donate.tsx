import i18n from 'assets/i18n';
import BigNumber from 'bignumber.js';
import {
    iptcColors,
    updateCommunityInfo,
    formatInputAmountToTransfer,
    getCurrencySymbol,
} from 'helpers/index';
import { ICommunityInfo, IRootState } from 'helpers/types';
import React, { Component } from 'react';
import { StyleSheet, Clipboard, Alert, View, TextInput } from 'react-native';
import {
    // Button,
    Paragraph,
    Portal,
    Snackbar,
    Text,
    Modal,
    Headline,
    IconButton,
    // Card,
} from 'react-native-paper';
import { ConnectedProps, connect } from 'react-redux';
import { analytics } from 'services/analytics';
import { celoWalletRequest } from 'services/celoWallet';
import { writeLog } from 'services/logger/write';
import config from '../../../../config';
import * as Device from 'expo-device';
import Button from 'components/Button';
import Card from 'components/Card';

interface IExploreScreenProps {
    community: ICommunityInfo;
}
const mapStateToProps = (state: IRootState) => {
    const { user, app } = state;
    return { user, app };
};

const connector = connect(mapStateToProps);
type PropsFromRedux = ConnectedProps<typeof connector>;
type Props = PropsFromRedux & IExploreScreenProps;

interface IDonateState {
    openModalDonate: boolean;
    donating: boolean;
    amountDonate: string;
    showCopiedToClipboard: boolean;
    modalConfirmSend: boolean;
    rates: any;
}
BigNumber.config({ EXPONENTIAL_AT: [-7, 30] });
class Donate extends Component<Props, IDonateState> {
    constructor(props: any) {
        super(props);
        this.state = {
            openModalDonate: false,
            donating: false,
            amountDonate: '',
            showCopiedToClipboard: false,
            modalConfirmSend: false,
            rates: this.props.app.exchangeRates,
        };
    }

    handleConfirmDonateWithCeloWallet = () => {
        const { amountDonate } = this.state;
        const amountInDollars =
            parseFloat(formatInputAmountToTransfer(amountDonate)) /
            this.props.user.user.exchangeRate;
        if (
            amountInDollars >
            new BigNumber(this.props.user.celoInfo.balance)
                .dividedBy(10 ** config.cUSDDecimals)
                .toNumber()
        ) {
            Alert.alert(
                i18n.t('donate'),
                i18n.t('donationBiggerThanBalance'),
                [{ text: i18n.t('close') }],
                { cancelable: true }
            );
            return;
        }
        const { community, user } = this.props;
        Alert.alert(
            i18n.t('donate'),
            i18n.t('donateConfirmMessage', {
                symbol: getCurrencySymbol(user.user.currency),
                amount: amountDonate.replace(/\B(?=(\d{3})+(?!\d))/g, ','),
                amountInDollars: amountInDollars.toFixed(2),
                to: community.name,
            }),
            [
                {
                    text: i18n.t('donate'),
                    onPress: () => this.donateWithCeloWallet(),
                },
                {
                    text: i18n.t('cancel'),
                },
            ],
            { cancelable: true }
        );
    };

    donateWithCeloWallet = async () => {
        this.setState({ donating: true });
        const stableToken = await this.props.app.kit.contracts.getStableToken();
        const cUSDDecimals = await stableToken.decimals();
        const amountInDollars =
            parseFloat(formatInputAmountToTransfer(this.state.amountDonate)) /
            this.props.user.user.exchangeRate;
        const txObject = stableToken.transfer(
            this.props.community.contractAddress,
            new BigNumber(amountInDollars)
                .multipliedBy(new BigNumber(10).pow(cUSDDecimals))
                .toString()
        ).txo;
        celoWalletRequest(
            this.props.user.celoInfo.address,
            stableToken.address,
            txObject,
            'donatetocommunity',
            this.props.app.kit
        )
            .then((tx) => {
                console.log(tx);
                // TODO: wait for tx confirmation and request UI update
                // update donated values
                setTimeout(
                    () =>
                        updateCommunityInfo(
                            this.props.user.celoInfo.address,
                            this.props
                        ),
                    10000
                );

                Alert.alert(
                    i18n.t('success'),
                    i18n.t('youHaveDonated'),
                    [{ text: 'OK' }],
                    { cancelable: false }
                );
                analytics('donate', { device: Device.brand, success: true });
            })
            .catch((e) => {
                writeLog({ action: 'donate', details: e.message });
                analytics('donate', { device: Device.brand, success: false });
                Alert.alert(
                    i18n.t('failure'),
                    i18n.t('errorDonating'),
                    [{ text: 'OK' }],
                    { cancelable: false }
                );
            })
            .finally(() => {
                this.setState({
                    openModalDonate: false,
                    donating: false,
                    amountDonate: '',
                });
            });
    };

    handleCopyAddressToClipboard = () => {
        Clipboard.setString(this.props.community.contractAddress);
        this.setState({ showCopiedToClipboard: true });
        setTimeout(() => this.setState({ showCopiedToClipboard: false }), 5000);
        this.setState({ openModalDonate: false });
    };

    render() {
        const {
            openModalDonate,
            donating,
            amountDonate,
            showCopiedToClipboard,
            rates,
        } = this.state;
        const { community, user } = this.props;

        const amountInDollars =
            parseFloat(formatInputAmountToTransfer(amountDonate)) /
            this.props.user.user.exchangeRate;
        const amountInCommunityCurrency =
            amountInDollars * rates[community.currency].rate;

        let donatingModalString = '';
        const backForDays =
            amountInDollars /
            new BigNumber(community.vars._claimAmount)
                .dividedBy(10 ** config.cUSDDecimals)
                .toNumber() /
            community.beneficiaries.added.length;
        if (backForDays < 1) {
            donatingModalString = i18n.t('yourDonationWillBack', {
                backNBeneficiaries: Math.max(
                    1,
                    Math.floor(
                        amountInDollars /
                            new BigNumber(community.vars._claimAmount)
                                .dividedBy(10 ** config.cUSDDecimals)
                                .toNumber()
                    )
                ),
            });
        } else {
            donatingModalString = i18n.t('yourDonationWillBackFor', {
                backNBeneficiaries: community.beneficiaries.added.length,
                backForDays: Math.floor(backForDays),
            });
        }

        const donateWithValoraButton =
            user.celoInfo.address.length > 0 ? (
                <Button
                    modeType="default"
                    bold={true}
                    loading={donating}
                    disabled={donating}
                    onPress={this.handleConfirmDonateWithCeloWallet}
                >
                    {i18n.t('donateWithValora')}
                </Button>
            ) : (
                <Button
                    icon="alert"
                    modeType="default"
                    bold={true}
                    style={{
                        backgroundColor: '#f0ad4e',
                    }}
                    onPress={() => {
                        Alert.alert(
                            i18n.t('failure'),
                            i18n.t('youAreNotConnected'),
                            [{ text: i18n.t('close') }],
                            { cancelable: false }
                        );
                    }}
                >
                    {i18n.t('donateWithValora')}
                </Button>
            );

        return (
            <>
                <Button
                    modeType="green"
                    bold={true}
                    style={styles.donate}
                    labelStyle={{
                        fontSize: 20,
                        lineHeight: 23,
                        color: 'white',
                        fontWeight: 'bold',
                        // backgroundColor: 'red',
                    }}
                    onPress={() => this.setState({ openModalDonate: true })}
                >
                    {i18n.t('donate')}
                </Button>
                <Snackbar
                    visible={showCopiedToClipboard}
                    onDismiss={() => {}}
                    action={{
                        label: '',
                        onPress: () => {
                            // Do something
                        },
                    }}
                >
                    {i18n.t('addressCopiedClipboard')}
                </Snackbar>
                <Portal>
                    <Modal
                        visible={openModalDonate}
                        onDismiss={() =>
                            this.setState({ openModalDonate: false })
                        }
                    >
                        <Card style={{ marginHorizontal: 20 }}>
                            <Card.Content>
                                <View style={{ height: 40 }}>
                                    <View
                                        style={{
                                            flex: 1,
                                            flexDirection: 'row',
                                            justifyContent: 'space-between',
                                        }}
                                    >
                                        <Headline style={{ marginVertical: 3 }}>
                                            {i18n.t('donateSymbol', {
                                                symbol: user.user.currency,
                                            })}
                                        </Headline>
                                        <IconButton
                                            icon="close"
                                            size={20}
                                            style={{ top: -5 }}
                                            onPress={() =>
                                                this.setState({
                                                    openModalDonate: false,
                                                })
                                            }
                                        />
                                    </View>
                                </View>
                                <View
                                    style={{
                                        backgroundColor: '#F6F6F7',
                                        // opacity: 0.27,
                                        alignItems: 'center',
                                        borderRadius: 5,
                                    }}
                                >
                                    <TextInput
                                        keyboardType="numeric"
                                        placeholder={i18n.t('amountSymbol', {
                                            symbol: getCurrencySymbol(
                                                user.user.currency
                                            ),
                                        })}
                                        style={{
                                            fontFamily: 'Gelion-Regular',
                                            fontStyle: 'normal',
                                            fontWeight: 'normal',
                                            fontSize: 40,
                                            lineHeight: 50,
                                            textAlign: 'center',
                                        }}
                                        value={amountDonate}
                                        onChangeText={(text) =>
                                            this.setState({
                                                amountDonate: text,
                                            })
                                        }
                                    />
                                    {amountDonate.length > 0 && (
                                        <Paragraph
                                            style={{ marginVertical: 10 }}
                                        >
                                            ~
                                            {`${getCurrencySymbol(
                                                community.currency
                                            )}${
                                                (Math.floor(
                                                    amountInCommunityCurrency *
                                                        100
                                                ) / 100).toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
                                            } ${community.currency}`}
                                        </Paragraph>
                                    )}
                                </View>
                                {amountDonate.length > 0 && (
                                    <Paragraph
                                        style={{
                                            marginBottom: 20,
                                            textAlign: 'center',
                                        }}
                                    >
                                        {donatingModalString}
                                    </Paragraph>
                                )}
                                <Button
                                    modeType="gray"
                                    bold={true}
                                    style={{
                                        // backgroundColor: '#F2F3F5',
                                        marginVertical: 10,
                                    }}
                                    onPress={this.handleCopyAddressToClipboard}
                                >
                                    {i18n.t('copyContractAddress')}
                                </Button>
                                {donateWithValoraButton}
                            </Card.Content>
                        </Card>
                    </Modal>
                </Portal>
            </>
        );
    }
}

const styles = StyleSheet.create({
    donate: {
        borderRadius: 0,
        height: 69,
    },
});

export default connector(Donate);
