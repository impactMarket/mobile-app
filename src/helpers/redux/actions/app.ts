import { ContractKit } from '@celo/contractkit';
import {
    appAction,
    SET_APP_FROM_WELCOME_SCREEN,
    SET_APP_SUSPECT_WRONG_DATETIME,
    SET_CELO_KIT,
    SET_EXCHANGE_RATES,
} from 'helpers/constants';
import { AppActionTypes } from 'helpers/types/redux';
import { Subscription } from '@unimodules/core';

export function setCeloKit(kit: ContractKit): AppActionTypes {
    return {
        type: SET_CELO_KIT,
        payload: kit,
    };
}

export function setAppSuspectWrongDateTime(
    suspect: boolean,
    timeDiff: number
): AppActionTypes {
    return {
        type: SET_APP_SUSPECT_WRONG_DATETIME,
        payload: {
            suspect,
            timeDiff,
        },
    };
}

export function SetAppFromWelcomeScreen(nextScreen: string): AppActionTypes {
    return {
        type: SET_APP_FROM_WELCOME_SCREEN,
        payload: nextScreen,
    };
}

export function setAppExchangeRatesAction(exchangeRates: any): AppActionTypes {
    return {
        type: SET_EXCHANGE_RATES,
        payload: exchangeRates,
    };
}

export function setPushNotificationListeners(notificationsListeners: {
    notificationReceivedListener: Subscription;
    notificationResponseReceivedListener: Subscription;
}): AppActionTypes {
    return {
        type: appAction.SET_PUSH_NOTIFICATION_LISTENERS,
        payload: notificationsListeners,
    };
}
