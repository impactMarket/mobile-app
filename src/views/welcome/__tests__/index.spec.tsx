import { useNavigation } from '@react-navigation/core';
import { assert } from 'chai';
import { shallow, ShallowWrapper } from 'enzyme';
import { Screens } from 'helpers/constants';
import { SetAppFromWelcomeScreen } from 'helpers/redux/actions/app';
import React, { useEffect } from 'react';
import { Provider, useDispatch } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import Welcome from '../index';

const mockStore = configureMockStore();
const store = mockStore({});

describe('Welcome screen test suite', () => {
    let screen: ShallowWrapper<any>;

    const navigate = jest.fn();
    const dispatch = jest.fn();

    beforeEach(() => {
        jest.resetAllMocks();
        (useEffect as jest.Mock).mockImplementation((f: any) => f());
        (useNavigation as jest.Mock).mockReturnValue({ navigate });
        (useDispatch as jest.Mock).mockReturnValue(dispatch);
    });

    it('should render Welcome screen correctly', () => {
        givenScreen();
        thenItRenderProperly();
    });

    it('should navigate to Valora', () => {
        givenScreen();
        thenItRenderProperly();
        whenPressingButton('connectWithValora');
        thenSuccessActionsAreDispatch('Auth');
    });

    it('should navigate to Communities', () => {
        givenScreen();
        thenItRenderProperly();
        whenPressingButton('exploreCommunities');
        thenSuccessActionsAreDispatch('Communities');
    });

    function givenScreen() {
        assert.isDefined(Welcome);

        screen = shallow(
            <Provider store={store}>
                <Welcome />
            </Provider>
        )
            .childAt(0)
            .dive();
    }

    function whenPressingButton(id: string) {
        const btn = screen.find(`[testID="${id}"]`);
        expect(btn).toBeDefined();
        btn.simulate('press');
    }

    function thenSuccessActionsAreDispatch(screen: string) {
        switch (screen) {
            case 'Auth':
                expect(dispatch).toHaveBeenNthCalledWith(
                    1,
                    SetAppFromWelcomeScreen(Screens.Auth)
                );
                break;
            case 'Communities':
                expect(dispatch).toHaveBeenNthCalledWith(
                    1,
                    SetAppFromWelcomeScreen(Screens.Communities)
                );
                break;
        }
    }

    function thenItRenderProperly() {
        expect(screen).toMatchSnapshot();
    }
});
