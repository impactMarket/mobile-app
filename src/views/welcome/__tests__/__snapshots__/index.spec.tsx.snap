// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome screen test suite should navigate to Communities 1`] = `ShallowWrapper {}`;

exports[`Welcome screen test suite should navigate to Valora 1`] = `ShallowWrapper {}`;

exports[`Welcome screen test suite should render Welcome screen correctly 1`] = `ShallowWrapper {}`;
